gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java

```diff
@@ -0,0 +1,144 @@
+  @Getter
+  @Setter
+  protected PathFilter pathFilter;
+  protected final Map<String, List<FileStatus>> idToFileStatuses;
+
+  private Optional<EventSubmitter> eventSubmitter;
+  private SourceState state;
+
+  public SubsetFilesCopyableDatasetFinder(FileSystem fs, Properties props)
+      throws IOException {
+    super(fs, props);
+    this.identifierPattern = Pattern.compile(props.getProperty(IDENTIFIER_PATTERN, DEFAULT_IDENTIFIER_PATTERN));
+    this.pathFilter =
+        new RegexPathFilter(props.getProperty(SUBSETFILES_REGEX_FILTER, DEFAULT_SUBSETFILES_REGEX_FILTER));
+    this.rootPath = PathUtils.deepestNonGlobPath(this.datasetPattern);
+    this.idToFileStatuses = new HashMap<>();
+  }
+
+  public SubsetFilesCopyableDatasetFinder(FileSystem fs, Properties props, EventSubmitter eventSubmitter)
+      throws IOException {
+    this(fs, props);
+    this.eventSubmitter = Optional.of(eventSubmitter);
+  }
+
+  public SubsetFilesCopyableDatasetFinder(FileSystem fs, Properties props, EventSubmitter eventSubmitter,
+      SourceState state)
+      throws IOException {
+    this(fs, props, eventSubmitter);
+    this.state = state;
+  }
+
+  @Override
+  public List<CopyableDataset> findDatasets()
+      throws IOException {
+    List<CopyableDataset> datasets = Lists.newArrayList();
+    FileStatus[] fileStatuss = this.getDatasetDirs();
+    for (FileStatus datasetRootDir : fileStatuss) {
+      datasets.addAll(this.generateDatasetsByIdentifier(datasetRootDir.getPath()));
+    }
+    return datasets;
+  }
+
+  public List<CopyableDataset> generateDatasetsByIdentifier(Path datasetRootDirPath)
+      throws IOException {
+    List<CopyableDataset> datasets = Lists.newArrayList();
+    FileStatus[] fileStatuses = fs.listStatus(datasetRootDirPath, this.getPathFilter());
+    for (FileStatus fileStatus : fileStatuses) {
+      Matcher result = this.identifierPattern.matcher(fileStatus.getPath().getName().toString());
+      if (result.find()) {
+        String id = result.group(1);
+        if (idToFileStatuses.containsKey(id)) {
+          log.info("Adding " + fileStatus.getPath() + " to " + id);
+          idToFileStatuses.get(id).add(fileStatus);
+        } else {
+          List<FileStatus> entry = new ArrayList<>();
+          entry.add(fileStatus);
+          log.info("Adding " + fileStatus.getPath() + " to " + id);
+          idToFileStatuses.put(id, entry);
+        }
+      }
+    }
+    for (String id : idToFileStatuses.keySet()) {
+      datasets.add(this.datasetAndPathWithIdentifier(datasetRootDirPath, id));
+    }
+    return datasets;
+  }
+
+  public CopyableDataset datasetAndPathWithIdentifier(Path path, String identifier)
+      throws IOException {
+    List<Object> args =
+        Lists.newArrayList(fs, path, props, identifier, idToFileStatuses.get(identifier), eventSubmitter, state);
+    try {
+      return GobblinConstructorUtils.invokeLongestConstructor(SubsetFilesCopyableDataset.class, args.toArray());
```

Comment:
You can just call `GobblinConstructorUtils.invokeLongestConstructor(SubsetFilesCopyableDataset.class, fs, path, props, identifier, idToFileStatuses.get(identifier), eventSubmitter, state)`